src/ucs/arch/x86_64/cpu.c

```diff
@@ -292,4 +326,35 @@ void ucs_cpu_init()
 #endif
 }
 
+
+ucs_status_t ucs_arch_get_cache_size(ucs_cpu_cache_type_t type, size_t *value)
+{
+    x86_cache_line_reg_info_t cache_info;
+    x86_cache_line_reg_info_t line_info;
+    uint32_t sets, edx;
+    int i;
+
+    for (i = 0; i < 32; i++) {
+        asm volatile("cpuid"
```

Comment:
I believe that leaf 4 should be only used when leaf 2 explicitly told so.  Look at the description of the tag 0xff.

Also, generally you should not use a leaf without checking that it is less or eq to the max supported leaf, but for leafs 2 and 4 this is a moot point.